src/riak_kv_sweeper.erl

```diff
@@ -0,0 +1,369 @@
+                do_sweep(Index, State2);
+            State2 ->
+                State2
+        end,
+    {reply, ok, State3};
+
+handle_call(status, _From, State) ->
+    {ok, {Participants, Sweeps}, State1} = riak_kv_sweeper_state:status(State),
+    {reply, {Participants , Sweeps}, State1};
+
+handle_call(stop_all_sweeps, _From, State) ->
+    {ok, Running, State1} = riak_kv_sweeper_state:stop_all_sweeps(State),
+    {reply, Running, State1};
+
+handle_call(stop, _From, State) ->
+    {ok, _Running, State1} = riak_kv_sweeper_state:stop_all_sweeps(State),
+    {stop, normal, ok, State1}.
+
+handle_cast({update_started_sweep, Index, ActiveParticipants, Estimate}, State) ->
+    State1 = riak_kv_sweeper_state:update_started_sweep(State, Index, ActiveParticipants, Estimate),
+    {noreply, State1};
+
+handle_cast({sweep_result, Index, Result}, State) ->
+    State1 = riak_kv_sweeper_state:update_finished_sweep(State, Index, Result),
+    {noreply, State1};
+
+handle_cast({update_progress, Index, SweptKeys}, State) ->
+    State1 = riak_kv_sweeper_state:update_progress(State, Index, SweptKeys),
+    {noreply, State1};
+
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+handle_info({test_tick, Ref, From}, State) ->
+    {noreply, State1} = handle_info(sweep_tick, State),
+    From ! {ok, Ref},
+    {noreply, State1};
+
+handle_info(sweep_tick, State) ->
+    schedule_sweep_tick(),
+    State3 =
+        case lists:member(riak_kv, riak_core_node_watcher:services(node())) of
+            true ->
+                State1 = riak_kv_sweeper_state:update_sweep_specs(State),
+                case riak_kv_sweeper_state:maybe_schedule_sweep(State1) of
+                    {ok, Index, State2} ->
+                        do_sweep(Index, State2);
+                    State2 ->
+                        State2
+                end;
+            false ->
+                State
+        end,
+    {noreply, State3};
+
+handle_info(Msg, State) ->
+    lager:error("riak_kv_sweeper received unexpected message ~p", [Msg]),
+    {noreply, State}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+code_change(_OldVsn, State, _Extra) ->
+    {ok, State}.
+
+%% ====================================================================
+%% Internal Functions
+%% ====================================================================
+-spec do_sweep(index(), riak_kv_sweeper_state:state()) ->
+    riak_kv_sweeper_state:state().
+do_sweep(Index, State) ->
+    %% Ask for estimate before we call get_active_participants, since
+    %% riak_kv_index_tree clears the trees if they are expired.
+    AAEEnabled = riak_kv_entropy_manager:enabled(),
+    EstimatedKeys = get_estimate_keys(Index, AAEEnabled, State),
+
+    State2 =
+        case riak_kv_sweeper_state:get_active_participants(State, Index) of
+            {ok, [], State1} ->
+                State1;
+            {ok, ActiveParticipants, State1} ->
+                ?MODULE:update_started_sweep(Index,
```

Comment:
Doing a `?MODULE:some_fun` call is dubious. It triggers code upgrade, which does not seem to be the wish here.

I think the `?MODULE:` should be dropped.